frontend/src/components/user-service/signup-page/SignupForm.tsx

```diff
@@ -0,0 +1,122 @@
+            }
+        }
+        setPasswordStrength(passwordStrengthSum);
+    }
+
+    const showDisplayedSignupMessage = (message : string, type : DisplayedMessageTypes) => {
+        setDisplayedSignupMessage({message : message, type : type});
+    }
+
+    const startSigningUp = () => {
+        if(username==="" || emailAddress === "" || password === "" || confirmPassword === "") {
+            showDisplayedSignupMessage("All fields cannot be empty.", DisplayedMessageTypes.Error);
+            return;
+        }
+        if(!isUsernameValid()){
+            showDisplayedSignupMessage("Invalid username.", DisplayedMessageTypes.Error);
+            return;
+        }
+        if(!isEmailAddressValid()){
+            showDisplayedSignupMessage("Invalid email address.", DisplayedMessageTypes.Error);
+            return;
+        }
+        if(!isPasswordValid()){
+            showDisplayedSignupMessage("Invalid password.", DisplayedMessageTypes.Error);
+            return;
+        }
+        if(confirmPassword !== password){
+            showDisplayedSignupMessage("Password and confirm password does not match each other.", DisplayedMessageTypes.Error);
+            return;
+        }
+        showDisplayedSignupMessage("Signing up...", DisplayedMessageTypes.Info);
+        // TODO: send a signup request to the backend - after the signup feature is implemented in the backend
+    }
+
+    return(
+        <>
+          <form onSubmit={evt => {evt.preventDefault(); startSigningUp();}} className="w-3/4">
+            <p className="font-bold mb-1">Username:</p>
+            <InputFieldWithTip placeholder="Your username" onChange={setUsername} type="text">
+              {!isUsernameValid() && (<div className="text-red-300">Invalid username.</div>)}
+              {tip_username}
+            </InputFieldWithTip>
+            <p className="font-bold mt-3 mb-1">Email Address:</p>
+            <InputFieldWithTip placeholder="Your email" onChange={setEmailAddress} type="text">
+              {!isEmailAddressValid() && (<div className="text-red-300">Invalid email address.</div>)}
+              {tip_email}
+            </InputFieldWithTip>
+            <p className="font-bold mt-3 mb-1">Password:</p>
+            <InputFieldWithTip placeholder="Your password" onChange={passwordInputFieldOnChangeHandler} type="password">
+             {!isPasswordValid() && (<div className="text-red-300">Invalid password.</div>)}
+              <p>Password strength:</p>
+              <div className="h-3 flex flex-row bg-gray-300">
+                {passwordStrength > 0 && (<div className="bg-red-500" style={{width: Math.min(passwordStrength, 33) + "%"}}/>)}
+                {passwordStrength > 33 && <div className="bg-yellow-500" style={{width: (Math.min(passwordStrength, 67) - 33) + "%"}}/>}
+                {passwordStrength > 66 && <div className="bg-green-500" style={{width: (Math.min(passwordStrength, 100) - 66) + "%"}}/>}
+              </div>
```

Comment:
interesting algorithm for passwordStrength but.... `1234567812345678123456781234567812345678` is considered "secure" as well

Reply:
For this I think it would be a tricky problem, as different people may consider "strong password" with different ways/aspect. For your `1234567812345678123456781234567812345678` I think it would be secure since it's really long (although its consecutive). Also your `1234567812345678123456781234567812345678` is not a valid password since it's missing upper case letter and lower case letter. As the users themselves should be more responsible on their own password's strength, for now I think my algorithm is sufficient for simple brute force hack (along with failed login attempt check and account locking). If you think this is still not enough, you can feel free to modify my algorithm after merging this pull request. Thanks

Reply:
i think just add on an additional check for unsafe password like this, so like instead of showing the password security thing perhaps have some kind of message saying that the criteria for strong password is not met

Reply:
while for valid passwords continue what you've been doing

Reply:
It already has a notification of "invalid password" if certain criteria of the password is not met, so I think adding a notification again is redundant. Since users should have their own judges on whether a password is strong or not (I mean, password like `12345678...` is really easy to be judged by the user as an unsafe password despite what my algorithm shows). So I think a user-friendly general and responsive password strength checker would be better (what I mean by "general" does not include the special cases such as your `12345678...`. Again, users should be responsible for their own password's strength at the first time, our UI just provide a reference to them but not a definite result)

Reply:
i think that i'll leave it as a point of discussion to the group later since i think your implementation should be ok as is, just that i got a few suggestions to change them to account for such extreme cases. stay tuned on tele later 😃